### ISSUE TITLE
cls_pw (class-imbalance weighting) has no effect on OBB training — v8OBBLoss omits the class_weights multiply

### ISSUE THREAD

**Issue by IMBCIT** [ultralytics/ultralytics#24841]
**Version:** ultralytics 8.4.37 · torch 2.11.0+cu130

### Summary

Setting `cls_pw` for an OBB task has no effect on training. The class weights are computed and logged at train start, the detection loss applies them, but `v8OBBLoss` reimplements the classification-loss term and never multiplies by `class_weights` — so the parameter is silently inert for `task=obb`.

### Evidence — the divergence is one line

In `ultralytics/utils/loss.py`:

`v8DetectionLoss` applies the weights:
```python
self.class_weights = getattr(model, "class_weights", None)
...
bce_loss = self.bce(pred_scores, target_scores.to(dtype))
if self.class_weights is not None:
    bce_loss *= self.class_weights          # ← applied
loss[1] = bce_loss.sum() / target_scores_sum
```

`v8OBBLoss` computes BCE without it:
```python
loss[1] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum
#         ← no class_weights multiply
```

`DetectionTrainer.set_class_weights()` still runs for OBB (it logs `Class weights: [...]`), so the user sees weighting "active" while it has zero effect on gradients.

### Reproduction (deterministic)

Train the same OBB dataset 3× with `cls_pw=0.0`, `0.5`, `1.0` and a fixed `seed`. Expected: 3 different models. Actual: **bit-identical** weights and metrics (we observed identical holdout mAP50/recall/precision to 4 decimals across all three).

### Impact

OBB datasets with class imbalance can't use the documented imbalance lever; rare classes stay under-learned while the knob misleads. (Our case: a ~14× instance spread across look-alike sibling classes.)

### Suggested fix

Mirror the detect path — apply `class_weights` to the OBB BCE term, guarded by the same `if self.class_weights is not None`. Happy to open a PR with that one-line change plus a regression test if that's welcome.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @IMBCIT, thank you for the clear bug report and for digging into the OBB loss path so carefully 🚀 This is an automated response to help move things along, and an Ultralytics engineer will also assist soon.

Since this appears to be a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) if possible. In particular, a small standalone script and the exact training command/config used for your deterministic `cls_pw=0.0`, `0.5`, and `1.0` comparison would help us verify behavior quickly. If you can also include the relevant logs showing the reported class weights and identical outputs/metrics, that would be very useful 🔍

We also recommend a quick visit to the [Docs](https://docs.ultralytics.com/) for reference, including [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Thanks for reporting and debugging.

We will address this

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24842: Apply class weights in OBB loss](https://github.com/ultralytics/ultralytics/pull/24842) by @Y-T-G.

This change may resolve the issue by updating the OBB classification loss to support optional class weights during training. In particular, the PR:
- adds an intermediate `bce_loss` step in `ultralytics/utils/loss.py`
- checks for `self.class_weights` before finalizing the classification loss
- applies per-class weights to the BCE classification loss when provided
- keeps the rest of the loss pipeline unchanged, including box loss behavior

This should help with imbalanced datasets by giving underrepresented classes a stronger training signal, while preserving existing behavior when class weights are not set.

Credit to @Y-T-G for the fix and to everyone who helped report, investigate, and discuss the issue.

If you'd like to test whether this resolves the problem, you can try the latest `main` branch now:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Or, if preferred, you can wait for the next official PyPI release.

If you’re able to test it, feedback on whether the merged changes resolve the issue would be very helpful.

Thanks 🙏 for reporting this. If the issue still persists after testing, please share the results and any additional details.
